### PR TITLE
Feature cobra cmd

### PR DESCRIPTION
**Pull request by thisisibrahimd**
#236

### CODE REVIEW

**Comment by thisisibrahimd** [comment]
you can run the commands as such:
go run main.go master --config <path to config>
go run main.go worker --config <path to config>

**Comment by L1ghtman2k** [comment]
I will do a bit of a thorough review tomorrow, but at a glance, there are a few more things we will need to account for if worker and master become subcommands of the single executable. Specifically, platforming package:

(docker-swarm)
https://github.com/ScoreTrak/ScoreTrak/blob/30304a78a6a3884b76152214493680d9413132cf/pkg/platform/docker/docker.go#L201

(kubernetes)
https://github.com/ScoreTrak/ScoreTrak/blob/30304a78a6a3884b76152214493680d9413132cf/pkg/platform/kubernetes/kubernetes.go#L61

(docker)
https://github.com/ScoreTrak/ScoreTrak/blob/30304a78a6a3884b76152214493680d9413132cf/pkg/platform/docker/docker.go#L228

can not be referencing ./worker as the executable

**Comment by L1ghtman2k** [comment]
Also, lets remove Copyright lines from .go files. I think the LICENSE file in the root directory is sufficient

**Comment by L1ghtman2k** [comment]
At a glance seems fine. I will test them out and fix the formatting once the platforming files are fixes.

Also go lint github action works by highlighting code at: https://github.com/ScoreTrak/ScoreTrak/pull/248/files (scroll down in file, you will see)

**Comment by thisisibrahimd** [comment]
> Also, lets remove Copyright lines from .go files. I think the LICENSE file in the root directory is sufficient

yeah, the copyright lines were generated by the cobra cli auto generator. That can be removed

**Review by L1ghtman2k — CHANGES_REQUESTED** [review]

File: cmd/worker.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		log.Println("worker called")
+
+		q, err := queue.NewWorkerQueue(C.Queue)
+		if err != nil {
+			log.Panicf("%v", err)
+		} else {
+			return
+		}
+
+		q.Receive()
```
[the remark this answers is not in the document]
Comment (thisisibrahimd) [reply cmd/worker.go:28]: Fixed!

[the remark this answers is not in the document]
Comment (thisisibrahimd) [reply cmd/worker.go:28]: I just removed unnecessary else statement.
https://github.com/ScoreTrak/ScoreTrak/pull/248/commits/4de764d334225ac574e5c06c8252f8c70deba197

File: cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	// PubSubConfig
+	viper.SetDefault("pubSubConfig.reportForceRefreshSeconds", 60)
+	viper.SetDefault("pubSubConfig.channelPrefix", "master")
+
+	// JWT Config
+	viper.SetDefault("jwt.secret", "changeme")
+	viper.SetDefault("jwt.timeoutInSeconds", 86400)
+
+	viper.SetEnvPrefix("ST") // set the environment prefix
+	viper.AutomaticEnv()     // read in environment variables that match
```
Comment (L1ghtman2k) [inline cmd/root.go:131]: How does this intergrate with configor at pkg/config/staticconf.go?

Comment (L1ghtman2k) [reply cmd/root.go:131]: right now the defaults exist both at pkg/config/staticconf.go, and cmd/root.go

Comment (L1ghtman2k) [reply cmd/root.go:131]: Resolved over slack

File: cmd/master.go [hunk 3]
```diff
@@ -60,14 +62,121 @@ import (
...
+	err := cont.Invoke(func(d *gorm.DB) {
+		db = d
+	})
+	if err != nil {
+		return err
+	}
+	var tm time.Time
+	res, err := db.Raw("SELECT current_timestamp;").Rows()
+	if err != nil {
+		panic(err)
```
Comment (L1ghtman2k) [inline cmd/master.go:156]: lets not panic. this error could be somewhat expected, instead, lets return it to the parent function:

return err

File: cmd/worker.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"log"
+
+	"github.com/ScoreTrak/ScoreTrak/pkg/queue"
+	"github.com/spf13/cobra"
+)
+
+// workerCmd represents the worker command
+var workerCmd = &cobra.Command{
```
Comment (L1ghtman2k) [inline cmd/worker.go:11]: How does worker ingest config?

Comment (L1ghtman2k) [reply cmd/worker.go:11]: ah, never mind, seems like init from root will run. Will need to test this

Comment (thisisibrahimd) [reply cmd/worker.go:11]: The worker command injests the config via the global var in the root.go. This is how viper recommends to pass the config

File: cmd/master.go (hunk 3 above)
Comment (thisisibrahimd) [reply cmd/master.go]: https://github.com/ScoreTrak/ScoreTrak/pull/248/commits/8bb96233d1e79eefda053e717b13329064d0513b

File: pkg/config/dynamicconf.go [hunk 5]
```diff
@@ -5,17 +5,16 @@ import (
...
 
 // DynamicConfig model is a set of columns describing the dynamicConfig of the scoring engine
 type DynamicConfig struct {
 	ID uint64 `json:"id,omitempty"`
 	// Describes how long each round unit takes to execute in seconds. This value shuold have a minimum value enforced (something like 20 seconds)
-	RoundDuration uint64 `json:"round_duration,omitempty" default:"60"`
+	RoundDuration uint64 `json:"round_duration,omitempty"`
 	// Enables or disables competition globally
-	Enabled *bool `json:"enabled,omitempty" default:"false" gorm:"not null;default:false"`
+	Enabled *bool `json:"enabled,omitempty" gorm:"not null;default:false"`
```
Comment (L1ghtman2k) [inline pkg/config/dynamicconf.go:17]: I can't find the default values set for RoundDuration in root of viper

Comment (thisisibrahimd) [reply pkg/config/dynamicconf.go:17]: Adding that now. It took me a while to understand that the application injest both static and dynamic at start and uploads the dynamic config

Comment (L1ghtman2k) [reply pkg/config/dynamicconf.go:17]: Yeah, that part might be confusing 😅. But yes, Dynamic config gets uploaded at the start of the app, if it isn't uploaded already. And if it is already there, then we just pull latest dynamic config. Dynamic config syncs its settings over DB, and the values are able to be manipulated, and synced across all instances

Comment (thisisibrahimd) [reply pkg/config/dynamicconf.go:17]: sounds like something etcd would be a better solution for. viper evens support etcd connection

**Review by L1ghtman2k — DISMISSED** [review]

**Comment by L1ghtman2k** [comment]
Neat job. Lets test this on a kind cluster with helm, and push it up. Good job!

**Comment by thisisibrahimd** [comment]
working on it 👍

**Comment by thisisibrahimd** [comment]
no longer crashes as https://github.com/ScoreTrak/ScoreTrak/blob/feature-cobra-cmd/pkg/runner/runner.go#L114

**Comment by L1ghtman2k** [comment]
yeah, i am testing it to rn

**Comment by L1ghtman2k** [comment]
After having a call, we decided that this pull request introduces too many changes, and it is too hard to keep track of everything. We are instead splitting this PR into two different PRs:
- Introduce Cobra + Viper, however, keep the config package exactly the same as it was before(functionality-wise). This way platform package will only need to change its commands, like specified in this comment: https://github.com/ScoreTrak/ScoreTrak/pull/248#issuecomment-973893146
- Refactor Config, and any other packages that would be affected by a refactor
- Misc PR that introduces unrelated things like license, etc.

**Pull request closed without merging**